
# Analyzing Used Car Listings on eBay Kleinanzeigen

We will be working on a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

    dateCrawled - When this ad was first crawled. All field-values are taken from this date.
    name - Name of the car.
    seller - Whether the seller is private or a dealer.
    offerType - The type of listing
    price - The price on the ad to sell the car.
    abtest - Whether the listing is included in an A/B test.
    vehicleType - The vehicle Type.
    yearOfRegistration - The year in which which year the car was first registered.
    gearbox - The transmission type.
    powerPS - The power of the car in PS.
    model - The car model name.
    kilometer - How many kilometers the car has driven.
    monthOfRegistration - The month in which which year the car was first registered.
    fuelType - What type of fuel the car uses.
    brand - The brand of the car.
    notRepairedDamage - If the car has a damage which is not yet repaired.
    dateCreated - The date on which the eBay listing was created.
    nrOfPictures - The number of pictures in the ad.
    postalCode - The postal code for the location of the vehicle.
    lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.


In [142]:
import pandas as pd
import numpy as np 

In [143]:
# Load the cars file 
cars = pd.read_csv('autos.csv',encoding = 'Latin-1')

In [144]:
cars

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [145]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [146]:
cars.describe(include = 'all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN



Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.


In [147]:
# Clean the columns 
col_names = {'dateCrawled':'date_crawled','offerType':'offer_type','vehicleType':'vehicle_type','yearOfRegistration':'registration_year',
             'monthOfRegistration':'registration_month','fuelType':'fuel_type',
            'notRepairedDamage':'unrepaired_damage','nrOfPictures':'nr_of_pictures','postalCode':'postal_code',
             'lastSeen':'last_seen',
            'dateCreated':'ad_created'}
print(cars.columns)
cars.rename(col_names,axis=1,inplace= True)
print(cars.columns)


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')
Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [148]:
cars.isnull().sum()

date_crawled             0
name                     0
seller                   0
offer_type               0
price                    0
abtest                   0
vehicle_type          5095
registration_year        0
gearbox               2680
powerPS                  0
model                 2758
odometer                 0
registration_month       0
fuel_type             4482
brand                    0
unrepaired_damage     9829
ad_created               0
nr_of_pictures           0
postal_code              0
last_seen                0
dtype: int64

In [149]:
cars.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN



Our initial observations:

    There are a number of text columns where all (or nearly all) of the values are the same:
        seller
        offer_type
    The nr_of_pictures column looks odd, we'll need to investigate this further.



In [150]:
cars['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

Looks like this column has 0 as value for all the row.  So will drop this column as well 

In [151]:
cars.drop(['nr_of_pictures','seller','offer_type'],axis=1,inplace= True)

In [152]:
#Convert the price column and odometerto int from Object 
cars['price']= cars['price'].str.replace('$','').str.replace(',','').astype(int)
cars['odometer']= cars['odometer'].str.replace('km','').str.replace(',','').astype(int)
cars.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
cars[['price','odometer_km']].head(5)

price  odometer_km
0   5000       150000
1   8500       150000
2   8990        70000
3   4350        70000
4   1350       150000

#  Coverting German words to English

Identify categorical data that uses german words, translate them and map the values to their english counterparts

In [153]:
# For unrepaired_damage column
print(cars[['unrepaired_damage']].value_counts())
cars['unrepaired_damage'] = cars['unrepaired_damage'].map({'nein':'No','ja':'Yes'})
print(cars[['unrepaired_damage']].value_counts())


unrepaired_damage
nein                 35232
ja                    4939
dtype: int64
unrepaired_damage
No                   35232
Yes                   4939
dtype: int64


In [155]:
# For gearbox column
print(cars[['gearbox']].value_counts())
cars['gearbox'] = cars['gearbox'].map({'manuell':'Manual','automatik':'Automatic'})
print(cars[['gearbox']].value_counts())


gearbox  
manuell      36993
automatik    10327
dtype: int64
gearbox  
Manual       36993
Automatic    10327
dtype: int64


## Exploring the price columns

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline
cars['price'].plot.box()
plt.show()

Looks like there are some ouliers in the Price ..lets explore and remove them 

In [25]:
print(cars['price'].value_counts().sort_index(ascending=False).head(10))
print(cars['price'].value_counts().sort_index(ascending=True).head(10))
print(cars['price'].describe())


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

There are a number of listings with prices below 30 dollars, including about 1,500 at zero dollars . There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is 1 dollar.  We will keep the $1 items, but remove anything above 350,000 $, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.


In [27]:
cars = cars[cars['price'].between(1,351000)]
cars['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring the Odometer column 

In [30]:
cars['odometer_km'].value_counts().sort_index(ascending= False)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64



We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.


# Exploring the date Columns


There are a number of columns with date information:

    date_crawled
    registration_month
    registration_year
    ad_created
    last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.


In [33]:
cars[['date_crawled','ad_created','last_seen','registration_month','registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   ad_created          48565 non-null  object
 2   last_seen           48565 non-null  object
 3   registration_month  48565 non-null  int64 
 4   registration_year   48565 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 2.2+ MB


In [34]:
cars[['date_crawled','ad_created','last_seen']].head(5)

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [37]:
# Explore the date crawled column
cars['date_crawled'].str[:11].value_counts(normalize=True, dropna = False).sort_index()

2016-03-05     0.025327
2016-03-06     0.014043
2016-03-07     0.036014
2016-03-08     0.033296
2016-03-09     0.033090
2016-03-10     0.032184
2016-03-11     0.032575
2016-03-12     0.036920
2016-03-13     0.015670
2016-03-14     0.036549
2016-03-15     0.034284
2016-03-16     0.029610
2016-03-17     0.031628
2016-03-18     0.012911
2016-03-19     0.034778
2016-03-20     0.037887
2016-03-21     0.037373
2016-03-22     0.032987
2016-03-23     0.032225
2016-03-24     0.029342
2016-03-25     0.031607
2016-03-26     0.032204
2016-03-27     0.031092
2016-03-28     0.034860
2016-03-29     0.034099
2016-03-30     0.033687
2016-03-31     0.031834
2016-04-01     0.033687
2016-04-02     0.035478
2016-04-03     0.038608
2016-04-04     0.036487
2016-04-05     0.013096
2016-04-06     0.003171
2016-04-07     0.001400
Name: date_crawled, dtype: float64

In [38]:
cars['date_crawled'].str[:11].value_counts(normalize=True, dropna = False).sort_values()

2016-04-07     0.001400
2016-04-06     0.003171
2016-03-18     0.012911
2016-04-05     0.013096
2016-03-06     0.014043
2016-03-13     0.015670
2016-03-05     0.025327
2016-03-24     0.029342
2016-03-16     0.029610
2016-03-27     0.031092
2016-03-25     0.031607
2016-03-17     0.031628
2016-03-31     0.031834
2016-03-10     0.032184
2016-03-26     0.032204
2016-03-23     0.032225
2016-03-11     0.032575
2016-03-22     0.032987
2016-03-09     0.033090
2016-03-08     0.033296
2016-04-01     0.033687
2016-03-30     0.033687
2016-03-29     0.034099
2016-03-15     0.034284
2016-03-19     0.034778
2016-03-28     0.034860
2016-04-02     0.035478
2016-03-07     0.036014
2016-04-04     0.036487
2016-03-14     0.036549
2016-03-12     0.036920
2016-03-21     0.037373
2016-03-20     0.037887
2016-04-03     0.038608
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [39]:
# Explore the date last_seen column
cars['last_seen'].str[:11].value_counts(normalize=True, dropna = False).sort_index()

2016-03-05     0.001071
2016-03-06     0.004324
2016-03-07     0.005395
2016-03-08     0.007413
2016-03-09     0.009595
2016-03-10     0.010666
2016-03-11     0.012375
2016-03-12     0.023783
2016-03-13     0.008895
2016-03-14     0.012602
2016-03-15     0.015876
2016-03-16     0.016452
2016-03-17     0.028086
2016-03-18     0.007351
2016-03-19     0.015834
2016-03-20     0.020653
2016-03-21     0.020632
2016-03-22     0.021373
2016-03-23     0.018532
2016-03-24     0.019767
2016-03-25     0.019211
2016-03-26     0.016802
2016-03-27     0.015649
2016-03-28     0.020859
2016-03-29     0.022341
2016-03-30     0.024771
2016-03-31     0.023783
2016-04-01     0.022794
2016-04-02     0.024915
2016-04-03     0.025203
2016-04-04     0.024483
2016-04-05     0.124761
2016-04-06     0.221806
2016-04-07     0.131947
Name: last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.


In [43]:
# Explore the ad_created  column
cars['ad_created'].str[:11].value_counts(normalize=True, dropna = False).sort_values()

2015-09-09     0.000021
2016-02-17     0.000021
2016-01-16     0.000021
2016-02-07     0.000021
2015-08-10     0.000021
                 ...   
2016-03-12     0.036755
2016-04-04     0.036858
2016-03-21     0.037579
2016-03-20     0.037949
2016-04-03     0.038855
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [45]:
# Explore the registartion year column 
cars["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64



The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.


# Dealing with Incorrect Registration Year Data
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [61]:
# Explore the % of cars which are not between 1900 and 2016
(~cars["registration_year"].between(1900,2016)).sum()/cars.shape[0]

0.038793369710697

Given that this is less than 4% of our data, we will remove these rows.


In [65]:
cars = cars[cars["registration_year"].between(1900,2016)]
cars["registration_year"].value_counts(normalize=True).sort_values(ascending = False).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.


# Exploring price by brand 

In [69]:
cars['brand'].value_counts(normalize= True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
Name: brand, dtype: float64



German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.


In [74]:
brand_counts =cars['brand'].value_counts(normalize= True)
common_brands = brand_counts[brand_counts>0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [92]:
brand_mean_price = {}
for brand in common_brands:
    brand_mean_price[brand] = int(cars[cars['brand'] == brand]['price'].mean())
    
brand_mean_price_sorted= sorted(brand_mean_price.items(),key= lambda kv:(kv[1],kv[0]),reverse = True)
brand_mean_price_sorted

[('audi', 9336),
 ('mercedes_benz', 8628),
 ('bmw', 8332),
 ('volkswagen', 5402),
 ('ford', 3749),
 ('opel', 2975)]

In [100]:
print(brand_mean_price)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749}




Of the top 5 brands, there is a distinct price gap:

   - Audi, BMW and Mercedes Benz are more expensive
   - Ford and Opel are less expensive
   - Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.



# Exploring Mileage 

In [95]:
brand_mean_mileage = {}
for brand in common_brands:
    brand_mean_mileage[brand] = int(cars[cars['brand'] == brand]['odometer_km'].mean())
    
brand_mean_mileage_sorted= sorted(brand_mean_price.items(),key= lambda kv:(kv[1],kv[0]),reverse = True)
brand_mean_mileage_sorted

[('audi', 9336),
 ('mercedes_benz', 8628),
 ('bmw', 8332),
 ('volkswagen', 5402),
 ('ford', 3749),
 ('opel', 2975)]

In [115]:
brands_price_series = pd.Series(brand_mean_price).sort_values(ascending= False)
brande_mileage_series = pd.Series(brand_mean_mileage).sort_values(ascending= False)

In [117]:
brands_price_mileage = pd.DataFrame(brande_mileage_series,columns= ['mean_mileage'])
brands_price_mileage['mean_price']= brands_price_series
brands_price_mileage

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749



The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.


# Most common brand/model combinations

In [162]:
cars[['brand','model']].value_counts().head(20)

brand          model      
volkswagen     golf           4024
bmw            3er            2761
volkswagen     polo           1757
opel           corsa          1735
               astra          1454
volkswagen     passat         1425
audi           a4             1291
bmw            5er            1183
mercedes_benz  c_klasse       1172
               e_klasse       1001
audi           a3              882
               a6              835
ford           focus           811
               fiesta          769
volkswagen     transporter     704
renault        twingo          677
peugeot        2_reihe         621
mercedes_benz  a_klasse        588
opel           vectra          574
smart          fortwo          574
dtype: int64

#  Cars price damage vs non-damaged

In [160]:

damaged_cars_mean_price = cars[cars['unrepaired_damage'] =='Yes']['price'].mean()
print("Damaged cars mean price is ", damaged_cars_mean_price)
non_damaged_cars_mean_price = cars[cars['unrepaired_damage'] =='No']['price'].mean()
print("Non Damaged cars mean price", non_damaged_cars_mean_price)

Damaged cars mean price 2311.889248835797
Non Damaged cars mean price 8470.59579359673


Non damaged cars are pricing almost 40% more than damaged cars